### PR TITLE
[WiP] Textfield to support RegExp validation

### PR DESCRIPTION
**Pull request by drehelis**
I was in a need for special form of `text` field validation done on the frontend. This PR tries to solve this by adding regular-expression validation capabilities.

Couple of caveats -
- Escaping characters such as `\` when editing JSON config manually is required.
- ~~Can't get the Admin UI properly generate `regex` object. Hence, WiP :-)~~
- Add Tests

Parameter config should look like:
```
{
  "name": "RegExp textfield",
  "max_length": "12",
  "regex": {
    "pattern": "^([a-z0-9_\\.-]+)@([\\da-z\\.-]+)\\.([a-z\\.]{2,6})$",
    "description": "Email address only"
  }
}
```
Small demo:

https://user-images.githubusercontent.com/11921235/121894690-1afb6100-cd28-11eb-8b36-a59df6a3c53f.mov

### CODE REVIEW

File: src/model/parameter_config.py
Comment (bugy) [inline src/model/parameter_config.py:285]: I think this is redundant

File: src/model/parameter_config.py
Comment (bugy) [inline src/model/parameter_config.py:300]: I think `bool` is not needed here

File: web-src/src/admin/components/scripts-config/ParameterConfigForm.vue
Comment (bugy) [inline web-src/src/admin/components/scripts-config/ParameterConfigForm.vue:76]: May be you could combine it with the max length row?

File: web-src/src/common/components/textfield.vue
Comment (bugy) [inline web-src/src/common/components/textfield.vue:201]: Since description can be empty, please provide a fallback value

File: web-src/src/common/components/textfield.vue
Comment (bugy) [inline web-src/src/common/components/textfield.vue:197]: Could you add a test to https://github.com/bugy/script-server/blob/master/web-src/tests/unit/textfield_test.js ?

File: src/model/parameter_config.py
Comment (bugy) [inline src/model/parameter_config.py:297]: Could you add a couple of tests?

File: src/model/parameter_config.py
Comment (bugy) [inline src/model/parameter_config.py:79]: Please also update the method "get_sorted_config" below

**Review by bugy — CHANGES_REQUESTED** [review]
Thanks, looks good to me, just a couple of improvement suggestions

File: web-src/src/common/components/textfield.vue
Comment (autogun) [reply web-src/src/common/components/textfield.vue:201]: When/if description is left empty, `regex.description` is the regex pattern itself. The idea is to provide the end-user with a meaningful error message when it comes to regex validation but if, for some reason, you don't, shouldn't print the regex pattern is enough of a fallback value?
https://github.com/bugy/script-server/blob/8ade772bb7218237b7ae741f895b7d960987c076/src/model/parameter_config.py#L302

File: web-src/src/admin/components/scripts-config/ParameterConfigForm.vue
Comment (autogun) [reply web-src/src/admin/components/scripts-config/ParameterConfigForm.vue:76]: You mean, move both `regex_pattern` & `regex_description` textfields under the same `div` with `max_length`?

Comment (bugy) [reply web-src/src/admin/components/scripts-config/ParameterConfigForm.vue:76]: yep

File: web-src/src/common/components/textfield.vue
Comment (bugy) [reply web-src/src/common/components/textfield.vue:201]: The python code you are referencing is a backend validation, a user should never see it. It's just an additional safety net.

File: web-src/src/common/components/textfield.vue
Comment (autogun) [reply web-src/src/common/components/textfield.vue:197]: How do I get started with tests? How do I run them locally on my dev env?
It would be nice to document this on the Wiki somewhere :-)

File: web-src/src/admin/components/scripts-config/ParameterConfigForm.vue
Comment (bugy) [inline web-src/src/admin/components/scripts-config/ParameterConfigForm.vue:252]: Please use camelCase for naming in javascript

**Review by bugy — CHANGES_REQUESTED** [review]

**Comment by ashishp0911** [comment]
Hello , 
This doesn't seem to work. I tried the exact pattern mentioned in this thread , but there doesn't seem to be any validation on the Input Field or am I missing something

The parameter section below : - 

  "parameters": [
    {
      "name": "Input",
      "required": true,
      "param": "-i",
      "env_var": "INP",
      "type": "text",
      "regex": {
        "pattern": "^([a-z0-9_\\.-]+)@([\\da-z\\.-]+)\\.([a-z\\.]{2,6})$",
        "description": "Input Validation"
      }
    }


Regards,
Ashish .A. Poojary

**Comment by bugy** [comment]
Hi @ashishp0911, how did you test it? This PR was not merged, i.e. not available in script server builds.

**Comment by ashishp0911** [comment]
@bugy My Bad , I spoke too early. I mistook the closed case #466 and thought this feature has been implemented in the latest build.
Any updates on the rollout of this feature ?

**Comment by bugy** [comment]
I have to apologize to @drehelis for missing that one. To be honest I was waiting for "review request", but I had to check this ticket on my own. Totally my fault for not tracking the project properly.

I made some minor adjustments, and will wait for build before merging. Probably tomorrow.

**Pull request merged**
